lib/http/server.js

```diff
@@ -0,0 +1,161 @@
+  StringPrototypeToLocaleLowerCase,
+} = primordials;
+
+const { cwd, nextTick } = process;
+const console = require('internal/console/global');
+const { createReadStream } = require('fs');
+const { Server } = require('_http_server');
+const { URL, isURLInstance, pathToFileURL } = require('internal/url');
+const { sep } = require('path');
+const { emitExperimentalWarning, kEmptyObject } = require('internal/util');
+const { getOptionValue } = require('internal/options');
+
+emitExperimentalWarning('http/server');
+
+if (module.isPreloading) {
+  const requiredModules = getOptionValue('--require');
+  if (ArrayPrototypeIncludes(requiredModules, `node:${module.id}`) ||
+      ArrayPrototypeIncludes(requiredModules, module.id)) {
+    const { parseArgs } = require('internal/util/parse_args/parse_args');
+    const { values: { port, host } } = parseArgs({ options: {
+      port: {
+        type: 'string',
+        short: 'p',
+      },
+      host: {
+        type: 'string',
+        short: 'h',
+      },
+    } });
+    nextTick(createStaticServer, { directory: process.argv[1], port, host });
+    process.env.NODE_REPL_EXTERNAL_MODULE = 'node:http/server';
+    process.argv[1] = 'node:http/server';
+  }
+}
+
+function createStaticServer(options = kEmptyObject) {
+  const {
+    directory = cwd(),
+    port,
+    host = 'localhost',
+    mimeOverrides = kEmptyObject,
+    serveDotFiles = false,
+  } = options;
+  const CWD_URL = isURLInstance(directory) ? directory : pathToFileURL(directory + sep);
+
+  const server = new Server(async (req, res) => {
+    const url = new URL(
+      req.url === '/' || StringPrototypeStartsWith(req.url, '/?') ?
+        './index.html' :
+        StringPrototypeReplace(new URL(req.url, 'root://').toString(), 'root://', '.'),
+      CWD_URL
+    );
+
+    if (
+      !serveDotFiles &&
+      StringPrototypeCharAt(
+        url.pathname,
+        StringPrototypeLastIndexOf(url.pathname, '/')
+      ) === '.'
+    ) {
+      console.log('401', req.url);
+      res.statusCode = 401;
+      res.end('Not Authorized');
+    }
+
+    const mime = {
+      '__proto__': null,
+      '.html': 'text/html; charset=UTF-8',
+      '.js': 'text/javascript; charset=UTF-8',
+      '.css': 'text/css; charset=UTF-8',
+      '.avif': 'image/avif',
+      '.svg': 'image/svg+xml',
+      '.jpg': 'image/jpeg',
+      '.jpeg': 'image/jpeg',
+      '.png': 'image/png',
+      '.webp': 'image/webp',
+      '.woff2': 'font/woff2',
+      ...mimeOverrides,
+    };
+    const ext = StringPrototypeToLocaleLowerCase(
+      StringPrototypeSlice(url.pathname, StringPrototypeLastIndexOf(url.pathname, '.'))
+    );
+    if (ext in mime) res.setHeader('Content-Type', mime[ext]);
+
+    try {
+      try {
+        const stream = createReadStream(url, { emitClose: false });
+        await new Promise((resolve, reject) => {
+          stream.on('error', reject);
+          stream.on('close', resolve);
+          stream.pipe(res);
+        });
```

Comment:
You should use `stream.pipeline` instead if this.

Reply:
`pipeline` implementation seems overly complicated in comparison of what's needed here (we don't need any `AbortSignal` support, etc.), and tries to do more that we want (i.e. it closes the request in case of error before we can send a 404 status).